File: test/integ/TypedefAnnoCheckerTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

import integ.TestIntDef;
import integ.TestStringDef;

@interface NotSafeAnno {}

interface I {
  int get();
}

public class TypedefAnnoCheckerTest {

  @TestIntDef int int_field;
  @TestStringDef int wrong_anno_field;
  @TestStringDef String str_field;

  void testIntField(@TestIntDef int val) {
    int_field = val;
  }

  void testWrongIntField(@TestIntDef int val) {
    wrong_anno_field = val;
  }

  void testStringField(@TestStringDef String val) {
    str_field = val;
  }

  static @NotSafeAnno @TestIntDef int testValidIntAnnoReturn(@NotSafeAnno @TestIntDef int val) {
    return val;
  }

  static @NotSafeAnno @TestStringDef String testValidStrAnnoReturn(@NotSafeAnno @TestStringDef String val) {
    return val;
  }

  static @TestIntDef int testIntAnnoInvokeStatic(@TestIntDef int val) {
    return testValidIntAnnoReturn(val);
  }

  static @TestStringDef String testStringAnnoInvokeStatic(@TestStringDef String val) {
    return testValidStrAnnoReturn(val);
  }

  static @TestIntDef String testWrongAnnotationReturned(@TestStringDef String val) {
    return val;
  }

  static @TestIntDef int testWrongAnnoInvokeStatic(@TestStringDef int val) {
    return testValidIntAnnoReturn(val);
  }

  static @NotSafeAnno @TestIntDef int testIrrelevantAnnos(@TestIntDef int val) {
    return val;
  }

  static @TestIntDef int testConstReturn() {
    @TestIntDef int val = TestIntDef.FOUR;
    return val;
  }

  static @TestIntDef int testInvalidConstReturn() {
    int val = 5;
    return val;
  }

  static @TestIntDef int testInvalidConstReturn2() {
    @TestIntDef int val = 5;
    return val;
  }

  static @TestStringDef String testConstStrReturn() {
   String val = "one";
    return val;
  }

  static @TestStringDef String testInvalidConstStrReturn() {
    @TestStringDef String val = "five";
    return val;
  }

  static @TestIntDef int testInvalidConstInvokeStatic() {
    int val = 5;
    return testIntAnnoInvokeStatic(val);
  }

  static @TestIntDef int testInvalidConstInvokeStatic2() {
    @TestIntDef int val = 5;
    return testIntAnnoInvokeStatic(val);
  }

  static @TestIntDef int testMultipleBlocksInt(@TestIntDef int val) {
    if (val > TestIntDef.TWO) {
      return val;
    }
    return TestIntDef.FOUR;
  }

  static @TestStringDef String testMultipleBlocksString(@TestStringDef String val) {
    if (val.equals(TestStringDef.THREE)) {
      return val;
    }
    return TestStringDef.ONE;
  }

  static @TestStringDef String testInvalidMultipleBlocksString(@TestStringDef String val) {
    if (val.equals(TestStringDef.THREE)) {
      val = val.concat("five");
      return val;
    }
    return TestStringDef.ONE;
  }

  static @TestIntDef int testNonConstInt(@TestIntDef int val) {
    val += 2;
    return val;
  }

  static @TestIntDef I testInvalidType(@TestIntDef I val) {
    return val;
  }

  static @TestStringDef String testJoiningTwoAnnotations(@TestStringDef String val, @TestIntDef String val2) {
    int flag = 1;
    @TestStringDef String s = val;
    if (flag == 1) {
      s = val;
    } else {
      s = val2;
    }
    return s;
  }

  static @TestStringDef String testJoiningTwoAnnotations2(@TestStringDef String val, @TestStringDef String val2) {
    int flag = 1;
    @TestStringDef String s = val;
    if (flag == 1) {
      s = val;
    } else {
      s = val2;
    }
    return s;
  }

  static @TestIntDef int testReassigningInt(@TestStringDef int val, @TestIntDef int val2) {
    val = val2;
    return val;
  }

  static @TestIntDef int testIfElse() {
    boolean flag = true;
    int res = flag ? TestIntDef.ONE : TestIntDef.ZERO;
    return res;
  }

  static @TestIntDef int testIfElseParam(boolean flag) {
    int res = flag ? TestIntDef.ONE : TestIntDef.ZERO;
    return res;
  }

  static @TestStringDef String testIfElseString(boolean flag) {
    String res = flag ? TestStringDef.ONE : TestStringDef.TWO;
    return res;
  }

  static @TestIntDef int testXORIfElse(boolean flag) {
    int res = flag ? TestIntDef.ZERO : TestIntDef.ONE;
    return res;
  }

  static @TestIntDef int testXORIfElseZero() {
    boolean flag = false;
    int res = flag ? TestIntDef.ZERO : TestIntDef.ONE;
    return res;
  }
}

File: test/integ/TypedefAnnoCheckerTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest

import integ.TestIntDef
import integ.TestStringDef

public class ClassWithParams(@TestIntDef var int_field: Int) {}

public class ClassWithDefaultParams(@TestIntDef var int_field: Int = TestIntDef.THREE) {}

public class TypedefAnnoCheckerKtTest {

  @TestStringDef val field_str: String = TestStringDef.TWO
  @TestStringDef var var_field: String = TestStringDef.THREE
  @NotSafeAnno val field_not_safe: String = "4"
  @TestIntDef var field_int: Int = TestIntDef.ONE

  companion object {
    @TestStringDef val companion_val: String = TestStringDef.ONE
    @TestStringDef var companion_var: String = TestStringDef.FOUR
    @TestIntDef var companion_int_var: Int = TestIntDef.FOUR
  }

  fun testSynthAccessor() {
    val lmd: () -> String = { takesStrConst("liu") }
    lmd()
  }

  private fun takesStrConst(@TestStringDef str: String): String {
    return str
  }

  fun wrongDefaultCaller(@TestStringDef arg: String) {
    wrongDefaultArg(arg)
    wrongDefaultArg()
  }

  private fun wrongDefaultArg(@TestStringDef str: String = "default"): String {
    return str
  }

  fun rightDefaultCaller(@TestStringDef arg: String) {
    rightDefaultArg(arg)
    rightDefaultArg()
  }

  private fun rightDefaultArg(@TestStringDef str: String = "one"): String {
    return str
  }

  @TestStringDef
  fun testCompanionObject(): String {
    return companion_val
  }

  @TestStringDef
  fun testCompanionVarSetter(): String {
    companion_var = TestStringDef.ONE
    return companion_var
  }

  @TestIntDef
  fun testCompanionIntVarSetter(): Int {
    companion_int_var = TestIntDef.ONE
    return companion_int_var
  }

  @TestStringDef
  fun testInvalidCompanionVarSetter(): String {
    companion_var = "5"
    return companion_var
  }

  @TestStringDef
  fun testKtField(): String {
    return field_str
  }

  @TestStringDef
  fun testVarField(): String {
    var_field = TestStringDef.ONE
    return var_field
  }

  @TestStringDef
  fun testInvalidVarField(): String {
    var_field = "5"
    return var_field
  }

  @TestIntDef
  fun testReturnWhen(): Int {
    return when (field_str) {
      "1" -> 1
      "2" -> 2
      "3" -> 3
      "4" -> 4
      else -> 0
    }
  }

  @TestStringDef
  fun call_lambda(lambda_func: () -> String): String {
    return lambda_func()
  }

  fun assign_and_print(@TestStringDef param: String) {
    var_field = param
    print(param)
  }

  fun assign_and_print_default(@TestStringDef param: String, def_param: String = "hello") {
    print(param)
    print(def_param)
  }

  @TestStringDef
  fun testLambdaCall(@TestStringDef param: String): String {
    return call_lambda({
      var_field = param
      assign_and_print(param)
      param
    })
  }

  @TestIntDef
  fun testClassConstructorArgs(@TestIntDef param: Int): Int {
    val ctor_test: ClassWithParams = ClassWithParams(param)
    return ctor_test.int_field
  }

  @TestIntDef
  fun testClassConstructorDefaultArgs(@TestIntDef param: Int): Int {
    val default_ctor_test: ClassWithDefaultParams = ClassWithDefaultParams(param)
    return default_ctor_test.int_field
  }

  @TestStringDef
  fun testLambdaCallLocalVar(): String {
    val local_val = "two"
    return call_lambda({
      assign_and_print_default(local_val)
      local_val
    })
  }

  @TestStringDef
  fun testLambdaCallLocalVarInvalid(): String {
    val local_val = "randomval"
    return call_lambda({
      assign_and_print_default(local_val)
      local_val
    })
  }

  ////////////////////////////////////////////////////////////////////////////////////////////
  interface Error {
    @get:TestStringDef val failure: String
  }

  abstract class BaseError : Error {}

  class Mismatch : BaseError() {
    @TestStringDef override val failure: String = TestStringDef.ONE
  }

  class Blowup : BaseError() {
    @TestStringDef override val failure: String = TestStringDef.THREE
  }

  fun getError(): Error {
    return Mismatch()
  }

  @TestStringDef
  fun testAnnotatedPropertyGetterPatching(): String {
    val e = getError()
    return e.failure
  }

  ////////////////////////////////////////////////////////////////////////////////////////////
  class Listener {
    companion object {
      @TestStringDef private val ONE: String = TestStringDef.ONE

      @TestStringDef
      fun getOne(): String {
        return ONE
      }
    }
  }

  // This test method is just a placehold to make the test look complete. The real testing target is
  // the getOne() method in the Companion object above.
  @TestStringDef
  fun testAnnotatedCompanionPropertyAccessorPatching(): String {
    return Listener.getOne()
  }
}
